File: sdk/typescript/src/sql.ts
import { RankingError } from './errors.js';
import { resolveConfig } from './presets.js';
import { tauSeconds } from './tuning.js';
import type { RankingConfig, RankingOptions } from './types.js';

/** SQL dialects with a tested emitter. */
export type Dialect = 'postgres' | 'mysql';

/**
 * `'B'` reproduces the library's power law exactly, over a bounded candidate set.
 * `'A'` swaps the power law for the exponential family so the ordering is fixed
 * and therefore indexable. See `toSQL()` for why that choice is forced.
 */
export type Strategy = 'A' | 'B';

export interface SQLOptions extends RankingOptions {
  /** Table to read from. Defaults to `'posts'`. */
  readonly table?: string;
  /** Signal field -> column name, when they differ. */
  readonly columns?: Readonly<Record<string, string>>;
  /** Timestamp column, used by strategy B. Defaults to `'created_at'`. */
  readonly dateColumn?: string;
  /** Immutable `bigint` epoch column, required by strategy A. Defaults to `'created_epoch'`. */
  readonly epochColumn?: string;
  /** Name of the generated column in strategy A. Defaults to `'hot'`. */
  readonly scoreColumn?: string;
  /** Candidate window for strategy B, in days. Defaults to 7. */
  readonly windowDays?: number;
  /** `LIMIT` in the emitted `SELECT`. Defaults to 20. */
  readonly limit?: number;
  /** Override the derived exponential time constant of strategy A. */
  readonly tauSeconds?: number;
  /** Seconds subtracted from the epoch to keep the numbers small. Defaults to 2020-01-01. */
  readonly epochOffset?: number;
  readonly dialect?: Dialect;
}

/** 2020-01-01T00:00:00Z. Counting from here rather than 1970 preserves float precision. */
export const EPOCH_OFFSET_2020 = 1_577_836_800;

const IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_$]*$/;

/** Validate a name and return it bare. Anything needing escaping is rejected. */
function identifier(name: string, label: string): string {
  if (!IDENTIFIER.test(name)) {
    throw new RankingError(
      `${label} ${JSON.stringify(name)} is not a plain SQL identifier. ` +
        'Pass a bare name matching [A-Za-z_][A-Za-z0-9_$]*, or write the SQL by hand.',
    );
  }
  return name;
}

/**
 * Quote an identifier for the target dialect.
 *
 * Everything is quoted, always. `like` — the default weight name in the
 * `social-feed` preset — is a reserved word in Postgres, so bare emission
 * produces a syntax error on the default path. Quoting also preserves the case of
 * camelCase columns, which is what ORM-created schemas usually have.
 */
function quoted(name: string, label: string, dialect: Dialect): string {
  const bare = identifier(name, label);
  return dialect === 'mysql' ? `\`${bare}\`` : `"${bare}"`;
}

/** Trim float noise so the emitted SQL stays readable. */
function num(value: number): string {
  const rounded = Number(value.toPrecision(12));
  return Number.isInteger(rounded) ? `${rounded}.0` : String(rounded);
}

function logFn(dialect: Dialect): string {
  // Postgres `log(x)` is already base 10; MySQL needs log10() since log() is natural.
  return dialect === 'postgres' ? 'log' : 'log10';
}

/** The `sum(w_i * p_i)` half, as a SQL expression. */
function engagementSQL(config: RankingConfig, options: SQLOptions, dialect: Dialect): string {
  if (config.engagement) {
    throw new RankingError(
      'A custom engagement() callback cannot be translated to SQL. ' +
        'Express the same thing as { signals } weights, or write the expression by hand.',
    );
  }
  const columns = options.columns ?? {};
  const terms: string[] = [];
  // Sorted, so the emitted SQL does not depend on how the signals object was
  // written — and so Go, whose maps have no iteration order, can match it.
  for (const field of Object.keys(config.signals).sort()) {
    const weight = config.signals[field]!;
    if (weight === 0) continue;
    const column = quoted(columns[field] ?? field, 'Column', dialect);
    terms.push(weight === 1 ? column : `${num(weight)} * ${column}`);
  }

  if (config.bayesian) {
    const { ratingField, countField, prior, priorCount, weight } = config.bayesian;
    const rating = quoted(columns[ratingField] ?? ratingField, 'Column', dialect);
    const count = quoted(columns[countField] ?? countField, 'Column', dialect);
    terms.push(
      `${num(weight)} * ((${num(priorCount)} * ${num(prior)} + ${count} * ${rating}) ` +
        `/ nullif(${num(priorCount)} + ${count}, 0))`,
    );
  }

  return terms.length === 0 ? '0' : terms.join(' + ');
}

/** Just the score expression, without the surrounding statement. */
export function sqlExpression(options: SQLOptions = {}, strategy: Strategy = 'B'): string {
  const config = resolveConfig(options);
  const dialect = options.dialect ?? 'postgres';
  const log = logFn(dialect);
  const sum = engagementSQL(config, options, dialect);

  if (strategy === 'A') {
    const epoch = quoted(options.epochColumn ?? 'created_epoch', 'Column', dialect);
    const tau = options.tauSeconds ?? tauSeconds(options);
    const offset = options.epochOffset ?? EPOCH_OFFSET_2020;
    if (!Number.isFinite(tau) || tau <= 0) {
      throw new RankingError(
        'Strategy A needs a finite, positive tau. With gravity = 0 time is ignored entirely, ' +
          'so there is nothing to index on — rank by engagement alone.',
      );
    }
    return `${log}(1 + ${sum}) + (${epoch} - ${num(offset)}) / ${num(tau)}`;
  }

  const dateColumn = quoted(options.dateColumn ?? 'created_at', 'Column', dialect);
  const age =
    dialect === 'postgres'
      ? `extract(epoch from now() - ${dateColumn}) / 3600.0`
      : `timestampdiff(second, ${dateColumn}, now()) / 3600.0`;
  return `${log}(1 + ${sum}) - ${num(config.gravity)} * ${log}((${age}) + ${num(config.graceHours)})`;
}

/**
 * Emit the SQL equivalent of a configuration.
 *
 * **The question that decides everything: does `now` cancel when comparing two
 * rows?** Under the power law time sits inside a log, so it does not cancel —
 * the order genuinely changes as the clock moves, and no static index can hold
 * it. Under the exponential family time enters linearly and `now` cancels
 * completely: only the difference of publication dates matters, and that never
 * changes. This is why Reddit's formula is linear in time, and it was not an
 * accident.
 *
 * - **Strategy A** (high volume): an indexed generated column. The exponential
 *   form, frozen order, constant-cost index scan, recomputed on a vote rather
 *   than every second. Recommended for `social-feed`.
 * - **Strategy B** (power law, bounded window): the exact formula, over a
 *   candidate set the `WHERE` clause has already cut to a few thousand rows.
 *   Valid whenever the relevance window is bounded — true of a feed, never of a
 *   catalogue. Recommended for `ecommerce` with a nightly batch recompute.
 */
export function toSQL(options: SQLOptions = {}, strategy: Strategy = 'B'): string {
  const dialect = options.dialect ?? 'postgres';
  if (dialect !== 'postgres' && dialect !== 'mysql') {
    throw new RankingError(`Unsupported dialect ${JSON.stringify(dialect)}. Supported: postgres, mysql.`);
  }
  const tableName = identifier(options.table ?? 'posts', 'Table');
  const table = quoted(tableName, 'Table', dialect);
  const limit = options.limit ?? 20;
  const expression = sqlExpression(options, strategy);

  if (strategy === 'A') {
    const scoreName = identifier(options.scoreColumn ?? 'hot', 'Column');
    const scoreColumn = quoted(scoreName, 'Column', dialect);
    const epochColumn = identifier(options.epochColumn ?? 'created_epoch', 'Column');
    const type = dialect === 'postgres' ? 'double precision' : 'double';
    return [
      `-- Strategy A: indexed generated column. Order is frozen, so the index stays valid.`,
      `-- ${epochColumn} must be a plain bigint written at insert time:`,
      `--   ${dialect === 'postgres' ? 'extract(epoch from created_at)' : 'unix_timestamp(created_at)'} is not immutable on a timestamp`,
      `--   with time zone, and generated columns must be immutable.`,
      `ALTER TABLE ${table} ADD COLUMN ${scoreColumn} ${type}`,
      `  GENERATED ALWAYS AS (`,
      `    ${expression}`,
      `  ) STORED;`,
      ``,
      `CREATE INDEX ${quoted(`${tableName}_${scoreName}_idx`, 'Index', dialect)} ON ${table} (${scoreColumn} DESC);`,
      ``,
      `SELECT * FROM ${table} ORDER BY ${scoreColumn} DESC LIMIT ${limit};`,
    ].join('\n');
  }

  const windowDays = options.windowDays ?? 7;
  if (!Number.isInteger(windowDays) || windowDays <= 0) {
    throw new RankingError(`windowDays must be a positive integer, received ${String(windowDays)}.`);
  }
  const dateColumn = quoted(options.dateColumn ?? 'created_at', 'Column', dialect);
  const window =
    dialect === 'postgres' ? `now() - interval '${windowDays} days'` : `now() - interval ${windowDays} day`;

  return [
    `-- Strategy B: exact power law over a bounded candidate set.`,
    `-- The WHERE clause uses an ordinary index on ${dateColumn} and cuts the set to a few`,
    `-- thousand rows; sorting that handful is free.`,
    `SELECT *,`,
    `       ${expression} AS score`,
    `FROM ${table}`,
    `WHERE ${dateColumn} > ${window}`,
    `ORDER BY score DESC`,
    `LIMIT ${limit};`,
  ].join('\n');
}

File: sdk/go/sql.go
package ranking

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

// EpochOffset2020 is 2020-01-01T00:00:00Z. Counting from here rather than 1970
// keeps the numbers small and preserves float precision.
const EpochOffset2020 = 1577836800

// Strategy selects which SQL translation to emit. See [ToSQL].
type Strategy string

const (
	// StrategyA is the indexed generated column: the exponential family, frozen
	// order, constant-cost index scan. Recommended for social-feed.
	StrategyA Strategy = "A"

	// StrategyB is the exact power law over a bounded candidate set.
	// Recommended for ecommerce with a nightly batch recompute.
	StrategyB Strategy = "B"
)

// SQLOptions configures the emitter. The zero value is usable: it emits Postgres
// against a table named "posts".
type SQLOptions struct {
	Table       string            // default "posts"
	Columns     map[string]string // signal name -> column name, when they differ
	DateColumn  string            // default "created_at"; strategy B
	EpochColumn string            // default "created_epoch"; strategy A
	ScoreColumn string            // default "hot"; strategy A
	WindowDays  int               // default 7; strategy B
	Limit       int               // default 20
	TauSeconds  float64           // default: derived from the config
	EpochOffset int64             // default EpochOffset2020
	Dialect     string            // "postgres" (default) or "mysql"
}

func (o SQLOptions) withDefaults() SQLOptions {
	if o.Table == "" {
		o.Table = "posts"
	}
	if o.DateColumn == "" {
		o.DateColumn = "created_at"
	}
	if o.EpochColumn == "" {
		o.EpochColumn = "created_epoch"
	}
	if o.ScoreColumn == "" {
		o.ScoreColumn = "hot"
	}
	if o.WindowDays == 0 {
		o.WindowDays = 7
	}
	if o.Limit == 0 {
		o.Limit = 20
	}
	if o.EpochOffset == 0 {
		o.EpochOffset = EpochOffset2020
	}
	if o.Dialect == "" {
		o.Dialect = "postgres"
	}
	return o
}

var identifierPattern = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_$]*$`)

// identifier validates a name and returns it bare. Anything needing escaping is
// rejected.
func identifier(name, label string) (string, error) {
	if !identifierPattern.MatchString(name) {
		return "", fmt.Errorf("%w: %s %q is not a plain SQL identifier; pass a bare name matching [A-Za-z_][A-Za-z0-9_$]*", ErrUnsupported, label, name)
	}
	return name, nil
}

// quoted quotes an identifier for the target dialect.
//
// Everything is quoted, always. "like" — the default weight name in the
// social-feed preset — is a reserved word in Postgres, so bare emission produces
// a syntax error on the default path. Quoting also preserves the case of
// camelCase columns, which is what ORM-created schemas usually have.
func quoted(name, label, dialect string) (string, error) {
	bare, err := identifier(name, label)
	if err != nil {
		return "", err
	}
	if dialect == "mysql" {
		return "`" + bare + "`", nil
	}
	return `"` + bare + `"`, nil
}

// num formats a float the way every SDK formats it, so the emitted SQL is
// byte-identical across TypeScript, JavaScript, Python and Go.
func num(value float64) string {
	rounded, err := strconv.ParseFloat(strconv.FormatFloat(value, 'g', 12, 64), 64)
	if err != nil {
		rounded = value
	}
	if rounded == math.Trunc(rounded) && math.Abs(rounded) < 1e15 {
		return strconv.FormatFloat(rounded, 'f', 1, 64)
	}
	return strconv.FormatFloat(rounded, 'g', -1, 64)
}

func logFn(dialect string) string {
	// Postgres log(x) is already base 10; MySQL needs log10() since log() is natural.
	if dialect == "postgres" {
		return "log"
	}
	return "log10"
}

func engagementSQL(cfg Config, opts SQLOptions) (string, error) {
	if cfg.Engagement != nil {
		return "", fmt.Errorf("%w: a custom Engagement func cannot be translated to SQL; express it as Signals weights, or write the expression by hand", ErrUnsupported)
	}
	var terms []string
	// Signal names are emitted in sorted order in every SDK, so the output does
	// not depend on how the map was built.
	for _, field := range sortedKeys(cfg.Signals) {
		weight := cfg.Signals[field]
		if weight == 0 {
			continue
		}
		name := field
		if mapped, ok := opts.Columns[field]; ok {
			name = mapped
		}
		column, err := quoted(name, "Column", opts.Dialect)
		if err != nil {
			return "", err
		}
		if weight == 1 {
			terms = append(terms, column)
		} else {
			terms = append(terms, num(weight)+" * "+column)
		}
	}

	if b := cfg.Bayesian; b != nil {
		ratingName, countName := b.RatingField, b.CountField
		if mapped, ok := opts.Columns[b.RatingField]; ok {
			ratingName = mapped
		}
		if mapped, ok := opts.Columns[b.CountField]; ok {
			countName = mapped
		}
		rating, err := quoted(ratingName, "Column", opts.Dialect)
		if err != nil {
			return "", err
		}
		count, err := quoted(countName, "Column", opts.Dialect)
		if err != nil {
			return "", err
		}
		terms = append(terms, fmt.Sprintf("%s * ((%s * %s + %s * %s) / nullif(%s + %s, 0))",
			num(b.Weight), num(b.PriorCount), num(b.Prior), count, rating, num(b.PriorCount), count))
	}

	if len(terms) == 0 {
		return "0", nil
	}
	return strings.Join(terms, " + "), nil
}

// SQLExpression returns just the score expression, without the surrounding
// statement.
func SQLExpression(cfg Config, strategy Strategy, opts SQLOptions) (string, error) {
	if err := cfg.Validate(); err != nil {
		return "", err
	}
	opts = opts.withDefaults()
	if opts.Dialect != "postgres" && opts.Dialect != "mysql" {
		return "", fmt.Errorf("%w: dialect %q; supported: postgres, mysql", ErrUnsupported, opts.Dialect)
	}
	logName := logFn(opts.Dialect)
	total, err := engagementSQL(cfg, opts)
	if err != nil {
		return "", err
	}

	if strategy == StrategyA {
		epoch, err := quoted(opts.EpochColumn, "Column", opts.Dialect)
		if err != nil {
			return "", err
		}
		tau := opts.TauSeconds
		if tau == 0 {
			if tau, err = TauSeconds(cfg); err != nil {
				return "", err
			}
		}
		if math.IsInf(tau, 0) || math.IsNaN(tau) || tau <= 0 {
			return "", fmt.Errorf("%w: strategy A needs a finite, positive tau; with gravity = 0 time is ignored entirely, so there is nothing to index on", ErrUnsupported)
		}
		return fmt.Sprintf("%s(1 + %s) + (%s - %s) / %s",
			logName, total, epoch, num(float64(opts.EpochOffset)), num(tau)), nil
	}

	date, err := quoted(opts.DateColumn, "Column", opts.Dialect)
	if err != nil {
		return "", err
	}
	age := fmt.Sprintf("extract(epoch from now() - %s) / 3600.0", date)
	if opts.Dialect == "mysql" {
		age = fmt.Sprintf("timestampdiff(second, %s, now()) / 3600.0", date)
	}
	return fmt.Sprintf("%s(1 + %s) - %s * %s((%s) + %s)",
		logName, total, num(cfg.Gravity), logName, age, num(cfg.GraceHours)), nil
}

// ToSQL emits the SQL equivalent of a configuration.
//
// # The question that decides everything
//
// Does now cancel when comparing two rows? Under the power law time sits inside
// a log, so it does not cancel: the order genuinely changes as the clock moves,
// and no static index can hold it. Under the exponential family time enters
// linearly and now cancels completely — only the difference of publication dates
// matters, and that never changes. This is why Reddit's formula is linear in
// time, and it was not an accident.
//
// StrategyA (high volume) is an indexed generated column: frozen order, constant
// cost, recomputed on a vote rather than every second.
//
// StrategyB (power law, bounded window) is the exact formula over a candidate set
// the WHERE clause has already cut to a few thousand rows. Valid whenever the
// relevance window is bounded — true of a feed, never of a catalogue.
func ToSQL(cfg Config, strategy Strategy, opts SQLOptions) (string, error) {
	opts = opts.withDefaults()
	expression, err := SQLExpression(cfg, strategy, opts)
	if err != nil {
		return "", err
	}
	tableName, err := identifier(opts.Table, "Table")
	if err != nil {
		return "", err
	}
	table, err := quoted(tableName, "Table", opts.Dialect)
	if err != nil {
		return "", err
	}

	if strategy == StrategyA {
		scoreName, err := identifier(opts.ScoreColumn, "Column")
		if err != nil {
			return "", err
		}
		scoreColumn, err := quoted(scoreName, "Column", opts.Dialect)
		if err != nil {
			return "", err
		}
		index, err := quoted(tableName+"_"+scoreName+"_idx", "Index", opts.Dialect)
		if err != nil {
			return "", err
		}
		epochColumn, err := identifier(opts.EpochColumn, "Column")
		if err != nil {
			return "", err
		}
		columnType, extractor := "double precision", "extract(epoch from created_at)"
		if opts.Dialect == "mysql" {
			columnType, extractor = "double", "unix_timestamp(created_at)"
		}
		return strings.Join([]string{
			"-- Strategy A: indexed generated column. Order is frozen, so the index stays valid.",
			fmt.Sprintf("-- %s must be a plain bigint written at insert time:", epochColumn),
			fmt.Sprintf("--   %s is not immutable on a timestamp", extractor),
			"--   with time zone, and generated columns must be immutable.",
			fmt.Sprintf("ALTER TABLE %s ADD COLUMN %s %s", table, scoreColumn, columnType),
			"  GENERATED ALWAYS AS (",
			"    " + expression,
			"  ) STORED;",
			"",
			fmt.Sprintf("CREATE INDEX %s ON %s (%s DESC);", index, table, scoreColumn),
			"",
			fmt.Sprintf("SELECT * FROM %s ORDER BY %s DESC LIMIT %d;", table, scoreColumn, opts.Limit),
		}, "\n"), nil
	}

	if opts.WindowDays <= 0 {
		return "", fmt.Errorf("%w: windowDays must be positive, got %d", ErrUnsupported, opts.WindowDays)
	}
	dateColumn, err := quoted(opts.DateColumn, "Column", opts.Dialect)
	if err != nil {
		return "", err
	}
	window := fmt.Sprintf("now() - interval '%d days'", opts.WindowDays)
	if opts.Dialect == "mysql" {
		window = fmt.Sprintf("now() - interval %d day", opts.WindowDays)
	}
	return strings.Join([]string{
		"-- Strategy B: exact power law over a bounded candidate set.",
		fmt.Sprintf("-- The WHERE clause uses an ordinary index on %s and cuts the set to a few", dateColumn),
		"-- thousand rows; sorting that handful is free.",
		"SELECT *,",
		"       " + expression + " AS score",
		"FROM " + table,
		fmt.Sprintf("WHERE %s > %s", dateColumn, window),
		"ORDER BY score DESC",
		fmt.Sprintf("LIMIT %d;", opts.Limit),
	}, "\n"), nil
}
